src/Symfony/Component/HttpKernel/Tests/Controller/ArgumentResolver/DateTimeValueResolverTest.php

```diff
@@ -0,0 +1,183 @@
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        $results = iterator_to_array($results);
+
+        $this->assertCount(1, $results);
+        $this->assertInstanceOf(\DateTime::class, $results[0]);
+        $this->assertEquals('2012-07-21', $results[0]->format('Y-m-d'));
+    }
+
+    public function testUnixTimestamp()
+    {
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTime::class, false, false, null);
+        $request = $this->requestWithAttributes(['dummy' => '989541720']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        $results = iterator_to_array($results);
+
+        $this->assertCount(1, $results);
+        $this->assertInstanceOf(\DateTime::class, $results[0]);
+        $this->assertEquals('2001-05-11', $results[0]->format('Y-m-d'));
+    }
+
+    public function testNullableWithEmptyAttribute()
+    {
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTime::class, false, false, null, true);
+        $request = $this->requestWithAttributes(['dummy' => '']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        $results = iterator_to_array($results);
+
+        $this->assertCount(1, $results);
+        $this->assertNull($results[0]);
+    }
+
+    public function testCustomClass()
+    {
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', FooDateTime::class, false, false, null);
+        $request = $this->requestWithAttributes(['dummy' => '2016-09-08 00:00:00']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        $results = iterator_to_array($results);
+
+        $this->assertCount(1, $results);
+        $this->assertInstanceOf(FooDateTime::class, $results[0]);
+        $this->assertEquals('2016-09-08', $results[0]->format('Y-m-d'));
+    }
+
+    public function testDateTimeImmutable()
+    {
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTimeImmutable::class, false, false, null);
+        $request = $this->requestWithAttributes(['dummy' => '2016-09-08 00:00:00']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        $results = iterator_to_array($results);
+
+        $this->assertCount(1, $results);
+        $this->assertInstanceOf(\DateTimeImmutable::class, $results[0]);
+        $this->assertEquals('2016-09-08', $results[0]->format('Y-m-d'));
+    }
+
+    public function testInvalidDate404Exception()
+    {
+        $this->expectException(NotFoundHttpException::class);
+        $this->expectExceptionMessage('Invalid date given for parameter "dummy".');
```

Comment:
I would move the `expectException...()` calls immediately before the code execution in which we expect the exception.

Same for the other tests

Reply:
Does it make sense to refactor `testInvalidDate404Exception`, `testAttributeWithFormatInvalidDate404Exception`, `testAttributeWithYmdFormatInvalidDate404Exception` to use a data provider? They look very similar

Reply:
Yes it does thanks :)